### PR TITLE
[Renderer] Use next, not return

### PR DESCRIPTION
**Pull request by NickLaMuro**
Fixes Jansa Failures: https://github.com/ManageIQ/manageiq-api/pull/933#issuecomment-715313395

In the context of this block, using `return` will eject from the iteration of the block, AND also the method, so in all of usages of `return` previously, it will `nil`, which is a valid return value for the method.

With `next`, it will just skip anything after the `yield` and move on to the next item in the iterator (which isn't obvious we are in one with this code change).  This is the behavior that was intended, but because the specs around this allow skipped at the proper times with the previous implementation, it wasn't know that this was a problem.

`next` is also used in the `determine_include_for_find` method, so this should be a relatively safe and tested change.

Links
-----

* Noticed in https://github.com/ManageIQ/manageiq-api/pull/933

### CODE REVIEW

**Comment by NickLaMuro** [comment]
@miq-bot add_label bug, jansa/yes?

This should be `jansa/yes`, as it is not only fixing a hidden bug on `master`, but also the failures currently on `jansa` described here:

https://github.com/ManageIQ/manageiq-api/pull/933#issuecomment-715313395

cc @jrafanie @Fryguy @kbrock

**Comment by Fryguy** [comment]
I guess this is sort of a side-effect of what I had concerns about here - https://github.com/ManageIQ/manageiq-api/pull/887#discussion_r485845138.  That is, the name of the method in no way implies that it's part of a looping construct, and so the caller doesn't know this effect will occur.  I wonder if it's better to make the block return nil instead of using next/return, thus making it feel more like a filter block a la `select`.

**Comment by NickLaMuro** [comment]
> I wonder if it's better to make the block return nil instead of using next/return, thus making it feel more like a filter block a la select.

I kind of agree, the only reason I am hesitant is that `determine_include_for_find` also uses `next`, so I would want to be consistent and change that as well, and I tend to prefer backporting only what is necessary.  I also like how he `next` style reads since you can more easily see what each of the conditions are at a glance, but agree clearly the legibility is causing bugs and obviously that is more important to avoid those.

Finishing a commit up now that fixes the tests as well (so they are a little less brittle), but let me know if you want the change from a `next` to a wrapping `if` statement.

**Review by kbrock — APPROVED** [review]
Great find nick.

Great tests that detect n+1 and  edge cases too.

**Comment by NickLaMuro** [comment]
Regarding the test failure:

I checked back on `master` and I got the same failure that I am getting here.  Even undid the changes from https://github.com/ManageIQ/manageiq-api/pull/933 and it was still present.

I think what changed is `ManageIQ/manageiq`, so I will have to do some digging to determine what failed there and is causing the issues here.

Sorry this is taking so long to fix... 😞

**Comment by NickLaMuro** [comment]
So after looking into master... I think the only PR(s) that could affect the output `spec/request/providers_spec.rb` was from here:

- https://github.com/ManageIQ/manageiq/pull/20725
- https://github.com/ManageIQ/manageiq-providers-openstack/pull/649

But I am not sure... I have commented everything from this and myself and can still not determine what is causing the issue...

**Comment by NickLaMuro** [comment]
Okay, after looking at this with a fresh set up eyes, turns out the culprit for the failures is not David, but @kbrock 

Test that is failing:

https://github.com/ManageIQ/manageiq-api/blob/b39c665878f6ae04b195d30cf5caeab05bb1cdcb/spec/requests/providers_spec.rb#L199-L232

PR responsible:

https://github.com/ManageIQ/manageiq/pull/20701

I unfortunately don't understand the code intent around this change, but this is affecting my ability to merge this PR, and thus fix the `jansa` which broke because of https://github.com/ManageIQ/manageiq-api/pull/933

Thanks.

cc @Fryguy @agrare

File: app/controllers/api/base_controller/renderer.rb [hunk 1]
```diff
@@ -609,9 +609,9 @@ def determine_include_for_find(klass)
 
       def determine_extra_cols(klass)
         virtual_attributes_for(klass) do |type, attr_name, attr_base|
-          return if attr_base.present?
-          return if virtual_attribute_accessor(type, attr_name)
-          return unless klass.attribute_supported_by_sql?(attr_name)
+          next if attr_base.present?
+          next if virtual_attribute_accessor(type, attr_name)
+          next unless klass.attribute_supported_by_sql?(attr_name)
```
Comment (jrafanie) [inline app/controllers/api/base_controller/renderer.rb:614]: wow, weird.  I don't follow what's happening here.  Is this method returning a collection of `attr_name.to_sym` that aren't skipped in the prior 3 conditionals?  Is it returning only the last `attr_name.to_sym`?  

I think this mistake happened because it's unclear it's looping through a list of attributes without an `each`.

**Comment by agrare** [comment]
@NickLaMuro I think this is a bug in how Openstack is building the child managers, it looks like it is calling `ensure_managers_zone_and_provider_region` before calling `ensure_cinder_manager` so the zone/region/tenant aren't being set properly.

I'll push a PR to openstack, thanks for tracking this down!

**Comment by NickLaMuro** [comment]
@agrare cool, thanks!

cc @kbrock (since I threw you under the bus for breaking this via https://github.com/ManageIQ/manageiq/pull/20701 )

**Comment by NickLaMuro** [comment]
@kbrock @jrafanie we good with this now?  Currently need this PR to "un-break" `jansa`

File: app/controllers/api/base_controller/renderer.rb (hunk 1 above)
Comment (jrafanie) [reply app/controllers/api/base_controller/renderer.rb:614]: This can be looked at later, it just looked odd to me so I thought it was worth mentioning

**Review by jrafanie — APPROVED** [review]
LGTM, my comment above can be looked at later.  For solving this bug, this PR is fine.

**Comment by simaishi** [comment]
Jansa backport details:

```
$ git log -1
commit 6dd642c080b1af8bc71f792af7420af6e54d3ef3
Author: Jason Frey <fryguy9@gmail.com>
Date:   Mon Oct 26 14:57:10 2020 -0400

    Merge pull request #940 from NickLaMuro/renderer-use-next-for-method-nested-blocks

    [Renderer] Use next, not return

    (cherry picked from commit d2990e2f94c78a7dff9d0696d66149da2f355287)
```

File: app/controllers/api/base_controller/renderer.rb (hunk 1 above)
Comment (NickLaMuro) [reply app/controllers/api/base_controller/renderer.rb:614]: @jrafanie while I am not going to "blame you" for why this looks like it does, but you definitely inspired the change:

https://github.com/ManageIQ/manageiq-api/pull/887#discussion_r481410812

Half :trollface:ing

Comment (jrafanie) [reply app/controllers/api/base_controller/renderer.rb:614]: Yeah, we can come back to it later.  It could be the wrong abstraction.  :trollface:
